qiskit/result/result.py

```diff
@@ -240,32 +240,35 @@ def _get_experiment(self, key=None):
             QiskitError: if there is no data for the experiment, or an unhandled
                 error occurred while fetching the data.
         """
-        if not self.success:
-            raise QiskitError(getattr(self, 'status',
-                                      'Result was not successful'))
-
         # Automatically return the first result if no key was provided.
         if key is None:
             if len(self.results) != 1:
                 raise QiskitError(
                     'You have to select a circuit or schedule when there is more than '
                     'one available')
-
             key = 0
 
-        # Key is an integer: return result by index.
-        if isinstance(key, int):
-            return self.results[key]
-
         # Key is a QuantumCircuit/Schedule or str: retrieve result by name.
         if isinstance(key, (QuantumCircuit, Schedule)):
             key = key.name
-
-        try:
-            # Look into `result[x].header.name` for the names.
-            return next(result for result in self.results
-                        if getattr(getattr(result, 'header', None),
-                                   'name', '') == key)
-        except StopIteration:
-            raise QiskitError('Data for experiment "%s" could not be found.' %
-                              key)
+        # Key is an integer: return result by index.
+        if isinstance(key, int):
+            exp = self.results[key]
+        else:
+            try:
+                # Look into `result[x].header.name` for the names.
+                exp = next(result for result in self.results
+                           if getattr(getattr(result, 'header', None),
+                                      'name', '') == key)
+            except StopIteration:
+                raise QiskitError('Data for experiment "%s" could not be found.' %
+                                  key)
+
+        # Check that the retrieved experiment was successful
+        if getattr(exp, 'success', False):
```

Comment:
As hinted via other channels, if the `Result` fully conforms to the schema, I think we can now count of `success` [being present](https://github.com/Qiskit/qiskit-terra/blob/master/qiskit/result/models.py#L49)? Not really blocking, though.